### PR TITLE
Feat/webhook auth

### LINKED ISSUE #2369: Add bearer/basic auth option to webhooks
Webhooks section needs optional bearer/basic authnetication

Image

### PR DESCRIPTION
**Pull request by harsh-aghara**
## Describe your changes

Add optional Basic Auth and Bearer Token authentication support for Webhook notifications.

**Server:**
- Added `WebhookAuthType` type definitions (`none` | `basic` | `bearer`)
- Extended Mongoose Notification schema with `authType`, `authUsername`, `authPassword`, `authToken` fields
- Added Zod validation with conditional requirements (username/password for Basic, token for Bearer)
- Updated `WebhookProvider.buildAuthHeaders()` to inject `Authorization` header into outgoing HTTP requests

**Client:**
- Added auth type selector dropdown (None / Basic Auth / Bearer Token) in webhook notification form
- Conditional form fields: username + password for Basic Auth, token for Bearer
- Client-side Zod validation with `superRefine` for conditional requirements
- Added all i18n translation keys, no hardcoded strings

## Write your issue number after "Fixes "

Fixes #2369

## Please ensure all items are checked off before requesting a review. "Checked off" means you need to add an "x" character between brackets so they turn into checkmarks.

- [x] (Do not skip this or your PR will be closed) I deployed the application locally.
- [x] (Do not skip this or your PR will be closed) I have performed a self-review and testing of my code.
- [x] I have included the issue # in the PR.
- [x] I have added i18n support to visible strings (instead of `<div>Add</div>`, use): 
```Javascript
const { t } = useTranslation();
<div>{t('add')}</div>
```
- [x] I have **not** included any files that are not related to my pull request, including package-lock and package-json if dependencies have not changed
- [x] I didn't use any hardcoded values (otherwise it will not scale, and will make it difficult to maintain consistency across the application).
- [x] I made sure font sizes, color choices etc are all referenced from the theme. I don't have any hardcoded dimensions.
- [x] My PR is granular and targeted to one specific feature.
- [x] I ran `npm run format` in server and client directories, which automatically formats your code.
- [x] I took a screenshot or a video and attached to this PR if there is a UI change.

None
image

Basic Auth
image

Bearer Token
image

### CODE REVIEW

File: client/src/Hooks/useNotificationForm.ts [hunk 1]
```diff
@@ -71,6 +75,9 @@ function buildDefaults(data: Notification | null): NotificationFormData {
 export const useNotificationForm = ({ data = null }: UseNotificationFormOptions = {}) => {
 	return useMemo(() => {
 		const defaults = buildDefaults(data);
-		return { schema: notificationSchema, defaults };
+		return {
+			schema: notificationSchema as unknown as typeof baseNotificationSchema,
```
Comment (ajhollid) [inline client/src/Hooks/useNotificationForm.ts]: This removes all type safety, please revert this

File: server/src/db/models/Notification.ts [hunk 2]
```diff
@@ -37,6 +37,14 @@ const NotificationSchema = new Schema<NotificationDocument>(
...
 		roomId: { type: String },
 		accessToken: { type: String },
+		authType: {
+			type: String,
+			enum: ["none", "basic", "bearer"] as WebhookAuthType[],
+			default: "none",
+		},
+		authUsername: { type: String },
+		authPassword: { type: String },
+		authToken: { type: String },
```
Comment (ajhollid) [inline server/src/db/models/Notification.ts:47]: These two fields should stripped before being returned to the front end, credentials should not be sent from the API to an unknown consumer.

File: server/src/db/models/Notification.ts [hunk 3]
```diff
@@ -37,6 +37,14 @@ const NotificationSchema = new Schema<NotificationDocument>(
 		homeserverUrl: { type: String },
 		roomId: { type: String },
 		accessToken: { type: String },
+		authType: {
+			type: String,
+			enum: ["none", "basic", "bearer"] as WebhookAuthType[],
```
Comment (ajhollid) [inline server/src/db/models/Notification.ts]: No casting please, use the correct type here

**Review by ajhollid — CHANGES_REQUESTED** [review]
In general, it looks decent.  There are some type casting issues to resolve, and credentials should be stripped so they are not returned from the API server.

File: server/src/db/models/Notification.ts (hunk 2 above)
Comment (harsh-aghara) [reply server/src/db/models/Notification.ts:47]: Reverted the type casts in the client hook and server schema. 
For security I've implemented sanitization at the Controller layer this ensures the WebhookProvider still has access to credentials for outgoing requests while stripping them from all API responses sent to the client.

File: server/src/controllers/notificationController.ts [hunk 4]
```diff
@@ -51,6 +51,15 @@ class NotificationController implements INotificationController {
...
 
+	private sanitizeNotification = (notification: any) => {
+		if (!notification) return notification;
+		const sanitized = { ...notification };
+		delete sanitized.authPassword;
+		delete sanitized.authToken;
+		delete sanitized.accessToken;
+		return sanitized;
+	};
+
```
Comment (ajhollid) [inline server/src/controllers/notificationController.ts:59]: Let's avoid using `delete`, we can just spread the properties and discard the unused ones

```
const {authPassword, authToken, accessToken, ...sanitized} = notification
return sanitized
```

That should be sufficient

**Review by ajhollid — CHANGES_REQUESTED** [review]
Thanks for making the requested changes, looking pretty good!  We can simplify that helper method to be more performant though, so let's do that.  Should be good to go after that!

File: server/src/controllers/notificationController.ts (hunk 4 above)
Comment (harsh-aghara) [reply server/src/controllers/notificationController.ts:59]: Done.
Thanks for the suggestion.

**Comment by pauln17** [comment]
Hey @harsh-aghara, I am working on an issue with slightly similar tasks, do you have a discord or something I could reach you on if you wouldn't mind?

**Comment by harsh-aghara** [comment]
> Hey @harsh-aghara, I am working on an issue with slightly similar tasks, do you have a discord or something I could reach you on if you wouldn't mind?

Sure.
You can mail me!

**Comment by harsh-aghara** [comment]
> Thanks for making the requested changes, looking pretty good!  We can simplify that helper method to be more performant though, so let's do that.  Should be good to go after that!

 @ajhollid   I have simplified the helper method.
Please review it.

**Comment by harsh-aghara** [comment]
@ajhollid Just pushed an update to clear out the branch rot, sync up with develop. 

What i changed:
* **Rebase & #3530:** Got everything rebased and made sure the new Pushover channel plays nicely with the shared validation schemas and UI.
* **Build fix:** Caught a TS type mismatch in the client that broke the build after recent merges. Fixed it and verified with a local npm run build.
* **Review feedback:** Swapped in the lighter helper method you suggested and added credential sanitization to the controller. 
* All unit and integration tests are passing locally again.

Could you approve the CI workflow when you have a sec and give it another look?

**Comment by harsh-aghara** [comment]
Hey @ajhollid, I just completed another rebase to resolve the fresh conflicts from the recently merged Twilio(#3534) and Pushover(#3530) PRs.

To make your review easier, here is what I've ensured in this update:

**Conflict Resolution:** Integrated the new Twilio/Pushover fields while keeping my Webhook Auth logic strictly isolated to avoid regressions.

**Maintainer Feedback:** Applied the performance-optimized sanitization (object destructuring) we discussed and ensured all credentials (including Webhook tokens) are stripped from API responses.

**Validation:** Verified both Client and Server builds locally, and all 995 tests are passing.

Since we were good to go after the last review, I'd love to get this merged before any more notification channels cause new conflicts. Let me know if there's anything else you need!

File: server/src/controllers/notificationController.ts [hunk 5]
```diff
@@ -37,16 +37,26 @@ class NotificationController implements INotificationController {
...
+				success: true,
+				msg: "Notification sent successfully",
 				details: { service: SERVICE_NAME },
 			});
 		} catch (error) {
 			next(error);
 		}
 	};
 
+	private sanitizeNotification = (notification: any) => {
```
Comment (Br0wnHammer) [inline server/src/controllers/notificationController.ts]: Nit - type this as `Notification` instead of any.

File: server/src/service/infrastructure/notificationProviders/webhook.ts [hunk 6]
```diff
@@ -6,6 +6,19 @@ import { getTestMessage } from "@/service/infrastructure/notificationProviders/u
 import got from "got";
 
 export class WebhookProvider extends NotificationProvider {
+	private buildAuthHeaders(notification: Partial<Notification>): Record<string, string> {
```
Comment (Br0wnHammer) [inline server/src/service/infrastructure/notificationProviders/webhook.ts:9]: There's already a `webhookProvider.test.ts` but nothing covers this. buildAuthHeaders is pure and security-sensitive, please add cases.

File: server/src/controllers/notificationController.ts [hunk 7]
```diff
@@ -37,16 +37,26 @@ class NotificationController implements INotificationController {
...
+		if (!notification) return notification;
+		const { authPassword, authToken, accessToken, ...sanitized } = notification;
```
Comment (Br0wnHammer) [inline server/src/controllers/notificationController.ts]: This strips `accessToken` from every notification, but that's the live credential for Telegram/Twilio/Pushover/Matrix too, won't this leave their token fields empty and block re-saving on edit?

**Review by Br0wnHammer — CHANGES_REQUESTED** [review]
Overall, a nice and clean implementation. The PR was easy to follow through. However, I have some questions and nitpicks. Please look at them.

File: server/src/controllers/notificationController.ts (hunk 5 above)
Comment (harsh-aghara) [reply server/src/controllers/notificationController.ts]: Done!
Typed it to Notification.

File: server/src/service/infrastructure/notificationProviders/webhook.ts (hunk 6 above)
Comment (harsh-aghara) [reply server/src/service/infrastructure/notificationProviders/webhook.ts:9]: Good call. I have added  test cases covering all the auth types and edge cases for missing credentials and empty strings.

File: server/src/controllers/notificationController.ts (hunk 7 above)
Comment (harsh-aghara) [reply server/src/controllers/notificationController.ts]: Ah I didn't realize that.
I've removed accessToken from the destructuring list so it's no longer stripped

File: server/src/controllers/notificationController.ts [hunk 8]
```diff
@@ -47,6 +48,12 @@ class NotificationController implements INotificationController {
 		}
 	};
 
+	private sanitizeNotification = (notification: Notification | null | undefined): Notification | null | undefined => {
+		if (!notification) return notification;
+		const { authPassword, authToken, ...sanitized } = notification;
```
Comment (Br0wnHammer) [inline server/src/controllers/notificationController.ts:54]: Since `sanitizeNotification` strips `authPassword/authToken` from responses, the edit form loads them empty, but `superRefine` still requires them, so editing a Basic/Bearer webhook seems to force re-entering credentials every time (and an empty submit would $set over the stored secret).

**Review by Br0wnHammer — CHANGES_REQUESTED** [review]

File: server/src/controllers/notificationController.ts (hunk 8 above)
Comment (harsh-aghara) [reply server/src/controllers/notificationController.ts:54]: I was focused on the sanitization that's why I didn't think through the "Edit" UX. It’s definitely a pain if a user has to re-enter their token just to rename a webhook. 
I can fix this by updating the Zod schema to allow empty strings on update and then tweak the service logic so it only $sets those fields if the user actually types something new. That way, leaving them blank would just keep the existing secret in the DB. Does that sound like the right path to you?

Comment (Br0wnHammer) [reply server/src/controllers/notificationController.ts:54]: @ajhollid what are your thoughts on this?

Comment (Br0wnHammer) [reply server/src/controllers/notificationController.ts:54]: @harsh-aghara, I was going through your proposed solution and I agree, leave blank to keep the secret is the right call. Just heads up that create and edit share the same validation, so loosening it would also let someone create a Basic auth webhook with no password. Probably want a separate lenient edit schema, and when authType changes, clear the old secret fields instead of skipping them so they don't linger in the DB.

Comment (harsh-aghara) [reply server/src/controllers/notificationController.ts:54]: @Br0wnHammer Thanks for the heads up.

I've just pushed a fix for this. I split out a lenient edit schema on both the client and server so users can safely leave secrets blank to keep their existing ones.

I also updated the service layer to explicitly $unset stale credentials if the authType (or the main channel type) changes, so we never leave lingering data in the DB.

Comment (Br0wnHammer) [reply server/src/controllers/notificationController.ts:54]: @harsh-aghara have your pushed the changes? I am not seeing any latest commits in the PR.

Comment (harsh-aghara) [reply server/src/controllers/notificationController.ts:54]: @Br0wnHammer Apologies for the confusion. I had to force-push to clean up an accidental commit that included unrelated files.

**Review by Br0wnHammer — APPROVED** [review]
LGMT! Great work on this @harsh-aghara 🚀

**Comment by Br0wnHammer** [comment]
Hi @harsh-aghara,
We discussed your PR with the team, and we've decided that we can't merge it in its current form due to a number of related changes that still need to be addressed.

Given the scope of the code changes, it would be best to open a new PR and incorporate the following feedback there:

* The edit schema is largely a duplicate of the create schema (roughly 90% overlap), so we'd like to avoid maintaining two nearly identical implementations.
* There are a few type casts that undermine type safety and should be refactored.
* It looks like notification testing may no longer work as expected, since some fields are being stripped before processing.

Once these concerns are addressed in a new PR, we can take another look.
Thankyou again for all your hardwork! 🚀
